### DISCUSSION TITLE
Investigate using native dialogs

### DISCUSSION THREAD

**Discussion by jandubois** [rancher-sandbox/rancher-desktop#5270] [Research]
On the Images tag, when we click "Add Images" we need to query the user for the image name.

Instead of displaying a dialog box, with a text entry field, and OK/Cancel[^1] buttons, we simulate a dialog in the window area:

[^1]: I'm not arguing about the button text, labeling them Cancel/Pull or whatever is fine, but conceptually they are "Abort operation" / "Perform operation" buttons.

![CleanShot 2023-08-01 at 12 19 55@2x](https://github.com/rancher-sandbox/rancher-desktop/assets/78947/9855a21f-4906-415f-a1f7-816ffc936d38)

It works like a non-modal dialog, but doesn't look like it at all. The Cancel button is in the top left corner (and labeled with a single `<` character), and the whole subwindow doesn't really look like a native dialog at all.

It also works differently from other non-modal dialogs in that you cannot navigate back to it once you click somewhere else (you can click on "Add Images" again, and get a new empty dialog).

The same design is being proposed for taking snapshots.

I would like to understand why we shouldn't switch to using regular dialog boxes here; that's what they have been designed for: when an action triggered by a button requires additional input from the user before it can be performed.

**Comment by kwwii** [comment]
This is an artefact of applying web UX to desktop apps. As mentioned in the last two sync calls, this behaviour needs to be investigated and properly defined.

**Reply by kwwii** [reply]
If other Rancher web apps will be installed/available as extensions how would this work?

**Reply by rak-phillip** [reply]
Other Rancher web apps that are installed through Rancher Desktop would behave just as they do today outside of Rancher Desktop. I don't think that we'd be able to do much to influence their behavior or design after we install them.

**Comment by kwwii** [comment]
Here is a simple mockup to using modals (based on the image list UI)
![modals](https://github.com/rancher-sandbox/rancher-desktop/assets/617368/d753ad95-54ba-4362-a8f3-2dfeaa1907de)

**Reply by rak-phillip** [reply]
I think there are two things to consider with using modal dialogs:

1. There's a dialog to unlock my OpenPGP key the first time I attempt to pull an image on Linux. This happens only if I configure Pass, as described in our installation instructions[^1]

![image](https://github.com/rancher-sandbox/rancher-desktop/assets/835961/00b512e0-0331-4acf-ae53-587e63e3a2f3)

2. There's also the "Console Output" section of this page that is rendered when pulling or building an image. We will need to understand where this belongs (if anywhere) if we switch to a dialog

![image](https://github.com/rancher-sandbox/rancher-desktop/assets/835961/6b75ce22-77f5-486b-9718-cc1c81e1ff7b)

[^1]:https://docs.rancherdesktop.io/getting-started/installation#pass-setup

**Comment by kwwii** [comment]
Using modals would require the modals to have a Save/Cancel action set(?)

**Reply by rak-phillip** [reply]
This is the design of our custom dialogs that we use today, they contain a title, detail, checkbox, and customizable buttons. Based on the mockups you provided, I think that Pull/Build would take the place of "Factory Reset" in the image below.

![image](https://github.com/rancher-sandbox/rancher-desktop/assets/835961/0e4cc7aa-f23b-4884-8964-3033ecf90cb9)

**Comment by rak-phillip** [comment]
I created issue #5284 for the "No Current Selection" highlighted in the image above. This is a bug that needs to be fixed regardless of the outcome of this discussion
